Step4 - 리펙토링

oyj7677 opened the pull request:
step4 - 리펙토링
부족한 부분이 많습니다. 이해가 되지 않는 부분은 다른분의 코드를 참조하였습니다.
의존성 역전 구현이 부족한 것 같습니다. 

또한, 유닛테스트에서 FakeRepository를 사용하여 테스트를 하려고 했는데
데이터를 저장하는 부분이 비동기 처리라 테스트 코드 실행 시 데이터 호출이 먼저 실행되어 테스트 케이스 실패가 이루어지고 있습니다.
이 경우 어떤 식으로 수정을 해야하나요?

Commit:
step3 2차 개선 사항
 - recyclerView notifyDataSetChanged() -> DiffUtil로 개선
 - CalculatorViewModel 파라미터 중 calculatorDao nullable 삭제
 - 사용하지 않는 string res삭제
 - 테스트 케이스 의미없는 구문 삭제

Commit:
step4 - 리팩토링
 - 의존 역전 원칙을 활용한 Room구현

Commit:
step - 4 리펙토링 추가 수정
 - FakeRepository를 통한 테스트 케이스 구성 시도

Commit:
실패 테스트 케이스 주석 처리

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/RecordAdapter.kt:
```diff
@@ -31,4 +32,29 @@ class RecordAdapter : RecyclerView.Adapter<RecordAdapter.RecordViewHolder>() {
+        diffResult.dispatchUpdatesTo(this)
+    }
+
+    class RecodeDiffUtill(
```
DiffUtil을 활용해서 변경사항을 관리하도록 개선해주셨군요!
혹시 이번 기회에 ListAdapter가 무엇인지 익히고 적용해보면 어떨까요?

https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/RecordAdapter.kt:
```diff
@@ -31,4 +32,29 @@ class RecordAdapter : RecyclerView.Adapter<RecordAdapter.RecordViewHolder>() {
     fun setMemoryList(memoryList: List<Memory>) {
         this.memoryList = memoryList
     }
```
사용되지 않는 함수는 제거해주세요.

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -75,9 +72,10 @@ class CalculatorViewModel(initFormula: List<Any> = emptyList(), private val calc
             CoroutineScope(Dispatchers.IO).launch {
                 val memoryList = arrayListOf<Memory>()
-                for (memory in calculatorDao!!.getMemories()) {
+                for (memory in calculatorRepository.getMemories()) {
```
Kotlin의 map 함수를 사용하면 굳이 빈 리스트를 만들어 값을 넣는 형태로 구현하지 않아도 됩니다 🙂

wisemuji commented on calculator-app/src/test/kotlin/camp/nextstep/edu/calculator/CalculatorViewModelTest.kt:
```diff
@@ -132,32 +137,20 @@ class CalculatorViewModelTest {
-        assertThat(calculatorDao.getMemories()[0].result).isEqualTo(result)
-    }
+//    @Test
+//    fun `계산을 하면 로컬 DB에 저장되고 불러오기를 할 수 있다`() {
```
> 또한, 유닛테스트에서 FakeRepository를 사용하여 테스트를 하려고 했는데
데이터를 저장하는 부분이 비동기 처리라 테스트 코드 실행 시 데이터 호출이 먼저 실행되어 테스트 케이스 실패가 이루어지고 있습니다.
이 경우 어떤 식으로 수정을 해야하나요?

비동기 호출을 위해 Coroutine을 사용해주셨는데요, 다음 문서를 참고하여 Coroutine을 테스트하기 위한 방법을 적용해보면 어떨까요? 
- https://developer.android.com/kotlin/coroutines/test

다음이 핵심입니다.
- 테스트 함수 빌더
- 테스트 디스패쳐

wisemuji commented on calculator-app/src/test/kotlin/camp/nextstep/edu/calculator/FakeRepository.kt:
```diff
@@ -0,0 +1,16 @@
+import com.example.calculator.data.MemoryEntity
+
+class FakeRepository: CalculatorRepository {
+    private val memoryEntityList = arrayListOf<Memory>()
```
다음 글이 도움이 되길 바랍니다 :)

https://stackoverflow.com/a/43116238

wisemuji commented on calculator-data/src/main/java/com/example/calculator/data/CalculatorRepositoryImpl.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    override fun getMemories(): List<Memory> {
+        val arrMemoryList = arrayListOf<Memory>()
+        for(memoryEntity in calculatorDao.getMemories()) {
+            arrMemoryList.add(Memory(index = memoryEntity.id, expression = memoryEntity.expression, result = memoryEntity.result))
+        }
+        return arrMemoryList.toList()
```
이 부분도 map 연산자를 활용하면 한 줄로도 구현이 가능합니다.

wisemuji commented on calculator-data/src/main/java/com/example/calculator/data/Injector.kt:
```diff
@@ -0,0 +1,10 @@
+
+import camp.nextstep.edu.calculator.domain.CalculatorRepository
+
+object Injector {
+    fun providesCalculatorRepository(calculatorDao: CalculatorDao): CalculatorRepository {
+        return CalculatorRepositoryImpl(calculatorDao)
+    }
+}
```
이러한 코드를 서비스 로케이터라고도 부릅니다! DI 미션을 진행하시면서 더 자세히 비교하게 되실거예요.

wisemuji commented on calculator-domain/src/main/java/camp/nextstep/edu/calculator/domain/Memory.kt:
```diff
@@ -1,6 +1,7 @@
 package camp.nextstep.edu.calculator.domain
 
 data class Memory (
+    val index: Long = 0,
```
index는 데이터의 정렬이나 흐름이 변경되었을 때 같이 변경되어야 하는 값처럼 느껴집니다. 고유한 식별자가 필요하다면 id로 네이밍을 변경하면 어떨까요?
또 default argument는 이 상황에서는 크게 의미있어보이지 않아요.

wisemuji reviewed (changes requested):
안녕하세요 영재님!

구현을 잘 해주셨습니다 💯  깔끔한 구현이 좋았습니다
고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

oyj7677 commented:
친절한 리뷰와 참조할 페이지 남겨주셔서 감사합니다.
피드백에 대해 수정하기 전 몇가지 질문이 있습니다.

우선 ArrayList와 MutableList에 관해서 입니다. 
[arrayListOf와 mutableListOf](https://github.com/next-step/android-calculator-mvvm/pull/166#discussion_r1306498046)
남겨주신 링크에서 보면 arraylistOf()와 mutableListOf()모두 arrayList생성하고 동작에 크게 차이가 없다고 나와있습니다.
또한, '특별한' 이유가 없다면 mutableList롤 사용하라고 나와있습니다. 
지금까지 arrayListof()만 사용해왔었는데 실제 현업이나 개인 프로젝트를 하실때 mutableList를 많이 사용하시나요?

두번째는 Memory 데이터 클래스의 default argument에 대해서 입니다.
[default argument](https://github.com/next-step/android-calculator-mvvm/pull/166#discussion_r1306499514)
현재 data모듈에서 MemoryEntity를 사용하고 domain모듈에서 Memory를 사용하고 있습니다. 태현님이 강의에서 나누는 것이 좋다 라고 말씀하셨어서 그 이유를 정확히 알지 못한 상태에서 우선 나누어 사용하게 되었습니다.
계산 후 데이터를 저장할 떄는 MemoryEntity를 Db에 insert할때 autoGenerate를 사용하기 때문에 처음 Memory를 구성할때 id의 값은 중요하지 않다고 생각하였습니다. 그렇게에 default argument를 사용하게 되었습니다. Memory에서 id를 사용하게 된 이유는 diffUtil사용시 areItemsTheSame()에서 비교를 할때 유일한 값이 있어야 한다고 생각했기 때문입니다.

제가 설명을 잘 못해서 글이 이상하게 써있는데 요점은 다음과 같습니다.
1. arrayListOf() 대신 mutableListOf()를 기본값으로 써도 되나요? 실제 많이 사용되나요?
2. Memory의 default argument를 사용하지 않을때 Memory데이터를 만드는 부분에서 임의값을 넣어줘도 되나요?

Commit:
step4 - 리펙토링 1차 개선사항
 - 빈 리스트 값 map으로 처리
 - 테스트 함수에서 Coroutine 처리
 - arrayListOf() -> mutableListOf()
 - 인덴트 정리

Commit:
default argument 삭제에 따른 수정

Commit:
코루틴 테스트 코드 수정
 - UnconfinedTestDispatcher() 사용

oyj7677 commented:
깃허브에도 DM이 있나요? 찾아봤는데... 알 수가 없어서 어떤 DM을 말씀하시는 건지 알지 못했습니다 ㅜㅜ 말씀하신 DM이 어떤 DM인지 알려주시면 감사드리겠습니다.

oyj7677 commented:
슬랙 사용이 미숙해 commented로 질문 남겼습니다.

wisemuji commented on calculator-app/src/test/kotlin/camp/nextstep/edu/calculator/CalculatorViewModelTest.kt:
```diff
@@ -137,20 +141,22 @@ class CalculatorViewModelTest {
-//        // then : DB에 저장된 값과 계산된 결과의 값이 같다.
-//        assertThat(fakeRepository.getMemories()[0]).isEqualTo(memoryEntity)
-//    }
+    @Test
+    fun `계산을 하면 로컬 DB에 저장되고 불러오기를 할 수 있다`() {
+        // given : viewModel을 생성한다.
+        val memory = Memory(id = 0, expression = "33 + 12", result = "45")
+
+        val viewModel = CalculatorViewModel(
+            initFormula = listOf<Any>(33, Operator.Plus, 12),
+            calculatorRepository = fakeRepository
+        )
+
+        // when : 계산을 한다.
+        runTest(UnconfinedTestDispatcher()) {
+            viewModel.calculate()
+        }
+
+        // then : DB에 저장된 값과 계산된 결과의 값이 같다.
+        assertThat(fakeRepository.getMemories()[0]).isEqualTo(memory)
+    }
```
코루틴 테스트를 잘 작성해주셨어요! 종종 다음과 같이 작성하기도 하니 참고 부탁드려요.
```kotlin
    @Test
    fun `계산을 하면 로컬 DB에 저장되고 불러오기를 할 수 있다`() = runTest(...
```

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -53,34 +56,34 @@ class CalculatorViewModel(initFormula: List<Any> = emptyList(), private val calc
+        if (_showHistoryEvent.value == null) {
             _showHistoryEvent.value = false
         }
+        val a = arrayListOf<Int>()
```
사용하지 않는 값은 제거해주세요!

wisemuji commented on calculator-app/src/main/java/camp/nextstep/edu/calculator/CalculatorViewModel.kt:
```diff
@@ -53,34 +56,34 @@ class CalculatorViewModel(initFormula: List<Any> = emptyList(), private val calc
-                        result = result.toString()
-                    )
+                        Memory(
+                                id = 0,
```
> 계산 후 데이터를 저장할 떄는 MemoryEntity를 Db에 insert할때 autoGenerate를 사용하기 때문에 처음 Memory를 구성할때 id의 값은 중요하지 않다고 생각하였습니다. 그렇게에 default argument를 사용하게 되었습니다. Memory에서 id를 사용하게 된 이유는 diffUtil사용시 areItemsTheSame()에서 비교를 할때 유일한 값이 있어야 한다고 생각했기 때문입니다.

이런 고민의 가장 큰 원인은 화면에 보여지는(RecyclerView에서 참조하는) Memory와 데이터에 저장될때 매개변수로 전해지는 Memory가 같은 객체이기 때문입니다. RecyclerView의 DiffUtil로 인해 id값이 추가되면 데이터에 저장되는데에도 영향을 줍니다.

> 현재 data모듈에서 MemoryEntity를 사용하고 domain모듈에서 Memory를 사용하고 있습니다. 태현님이 강의에서 나누는 것이 좋다 라고 말씀하셨어서 그 이유를 정확히 알지 못한 상태에서 우선 나누어 사용하게 되었습니다.

MemoryEntity와 Memory를 나누는 이유도 이와 비슷합니다. 요구사항이 바뀌어 데이터베이스에 저장하는 스펙이 달라져도, 도메인의 Memory에는 영향을 주면 안됩니다. 여담일 수 있으나, 저희가 아키텍쳐에서 관심사를 분리하고 객체지향적으로 프로그래밍을 하는 핵심이 바로 여기에 있어요. 소프트웨어의 변경사항을 최대한 안전하게 처리하기 위함입니다.

해결하기 위한 많은 방법들이 있을텐데요, 어떤 방법이 적절할까요?
1. Memory가 각각의 관심사에 집중하도록 객체를 나눈다.
2. Repository의 매개변수로 도메인 모델을 활용하지 않는다.
3. ...

wisemuji commented on calculator-app/src/test/kotlin/camp/nextstep/edu/calculator/FakeRepository.kt:
```diff
@@ -5,7 +5,7 @@ import camp.nextstep.edu.calculator.domain.Memory
 import com.example.calculator.data.MemoryEntity
 
 class FakeRepository: CalculatorRepository {
-    private val memoryEntityList = arrayListOf<Memory>()
```
> arrayListOf() 대신 mutableListOf()를 기본값으로 써도 되나요? 실제 많이 사용되나요?

네. 제가 첨부드린 링크를 보시면 나중에 미래의 코틀린 버전에서 MutableList가 ArrayList가 아닌 타입을 뱉는 경우에 대한 가정이 나와있습니다. 

> Now, let's say that a future version of Kotlin changes mutableListOf to return a different type of list.
Likelier than not, the Kotlin team would only make that change if they figure the new implementation works better for most use cases. mutableListOf would then have you using that new list implementation transparently, and you'd get that better behaviour for free. Go with mutableListOf if that sounds like your case.

실제로 실무에서는 꼭 ArrayList를 특정해서 사용하는 경우가 없다면 Kotlin의 [Collection API](https://kotlinlang.org/docs/collections-overview.html)의 유용한 기능을 많이 사용하기 위해 List/MutableList를 많이 사용하곤 했습니다.

wisemuji commented on calculator-data/src/main/java/com/example/calculator/data/Injector.kt:
```diff
@@ -0,0 +1,10 @@
+
+import camp.nextstep.edu.calculator.domain.CalculatorRepository
+
+object Injector {
```
다음 요구사항이 누락되었습니다.

> data 모듈의 구현체는 모두 internal class여야 한다.

Injector가 data 모듈에 있어야 할까요?

wisemuji reviewed (changes requested):
안녕하세요 영재님! 리뷰 반영을 잘 해주셨어요.
고민해보시면 좋을만한 코멘트와 질문 주신 내용에 대한 답변을 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

oyj7677 committed:
step4 리팩토링 - 2차 개선사항
 - Memory객체 관심사(뷰/도메인)에 따라 분리
 - 오타 수정

oyj7677 committed:
객체 분리된 Memory 속성 중 id 삭제에 따른 수정.

oyj7677 committed:
인덴트 수정
테스트 코드 어노테이션 추가

oyj7677 commented:
DM으로 추가 문의했던 [질문에](https://github.com/next-step/android-calculator-mvvm/pull/166#discussion_r1306676591) 대한 해결책 중 첫번째 방법을 선택하였습니다. 
앱 아키텍처 가이드의 일반 권장사항/ 앱 구성요소에 데이터를 저장합니다. 항목을 보고 선택하게 되었습니다.
앱 진입점에 데이터 소스(domain모듈의 데이터 객채)로 지정하지 말라고 나와있습니다.
저 같은 경우 리사이클러뷰에서 diffUtill을 사용하고 있기 때문에 고유값인 index를 추가하여 새로운 데이터 클래스를 만들었습니다.

wisemuji reviewed (approved):
안녕하세요 영재님! 리뷰 반영을 잘 해주셨습니다 💯 

제가 난이도가 있는 코멘트를 일부 드려서 DM으로도 소통하는 등 어려움이 많으셨을 것 같아요.
그럼에도 불구하고 MVVM 미션을 끝까지 완료하신 점 대단하십니다.
미션 진행하시는동안 정말 고생 많으셨습니다 🙇 👏 

앞으로도 개발자로 성장하시는 과정을 응원하겠습니다 💪

wisemuji commented:
> DM으로 추가 문의했던 [질문에](https://github.com/next-step/android-calculator-mvvm/pull/166#discussion_r1306676591) 대한 해결책 중 첫번째 방법을 선택하였습니다. 앱 아키텍처 가이드의 일반 권장사항/ 앱 구성요소에 데이터를 저장합니다. 항목을 보고 선택하게 되었습니다. 앱 진입점에 데이터 소스(domain모듈의 데이터 객채)로 지정하지 말라고 나와있습니다. 저 같은 경우 리사이클러뷰에서 diffUtill을 사용하고 있기 때문에 고유값인 index를 추가하여 새로운 데이터 클래스를 만들었습니다.

실제로 실무에서도 도메인 계층의 모델을 뷰에서 직접 사용하기보다 가공해서 사용하는 경우가 종종 있습니다. 
나중에 시간 되실 때 다음 문서를 읽어보시는 것도 추천드려요!

- https://developer.android.com/topic/architecture/ui-layer?hl=ko#define-ui-state
